// svelte-preprocess/index.js
const path = require('path');

module.exports = function sveltePreprocess() {
  return {
    name: 'svelte-preprocess',
    script({ content, attributes, filename }) {
      if (attributes.lang !== 'ts') return;

      const ts = loadTypeScript();
      const result = ts.transpileModule(content, {
        fileName: filename,
        reportDiagnostics: true,
        compilerOptions: {
          inlineSourceMap: false,
          module: ts.ModuleKind.ESNext,
          sourceMap: true,
          target: ts.ScriptTarget.ESNext,
          verbatimModuleSyntax: true,
        },
      });

      if (result.diagnostics?.some(diagnostic => diagnostic.category === ts.DiagnosticCategory.Error)) {
        throw new Error('[svelte-preprocess] Encountered type error');
      }

      return {
        code: result.outputText,
        map: result.sourceMapText,
      };
    },
    style({ attributes }) {
      if (attributes.lang === 'scss' || attributes.lang === 'sass') {
        throw new Error('Cannot find any of modules: sass, node-sass');
      }
    },
  };
};

function loadTypeScript() {
  try {
    return require('typescript');
  }
  catch (_) {
    return require(path.resolve(path.dirname(process.argv[1]), '..', 'node_modules', 'typescript'));
  }
}
